### PR TITLE
_posts: add 2023-01-05-mount-propagation-bug.md

### PR DESCRIPTION
**Pull request by brauner**
Signed-off-by: Christian Brauner (Microsoft) <christian@brauner.io>

### CODE REVIEW

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+The bug could be reproduced, albeit unreliably, by running the mount propagation test of the [LTP](https://github.com/linux-test-project/ltp) testsuite in a loop while simultaneously creating various network namespaces with `ip netns` command.
+When we started debugging this it turned out that the interesting aspect of `ip netns` for this bug was that it persisted a network namespace by bind-mounting it in a separate mount namespace.
+
+It turned out that the reliability of the reproducer could be significantly increased by using
+
+```
+unshare --mount --propagation=unchanged -- mount --make-rslave /
+```
+
+while using the [a specific test](https://github.com/linux-test-project/ltp/blob/af98698067f706feeb1729e038eef9aefc12760c/testcases/kernel/fs/fs_bind/bind/fs_bind24.sh#L4) of the LTP mount propagation testsuite modifying it slightly so that we loop around `mount` and `umount` in the script.
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]:
```suggestion
while using the [a specific test](https://github.com/linux-test-project/ltp/blob/af98698067f706feeb1729e038eef9aefc12760c/testcases/kernel/fs/fs_bind/bind/fs_bind24.sh#L4) of the LTP mount propagation testsuite, modifying it slightly so that we loop around `mount` and `umount` in the script.
```

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+
+In previous years Seth Forshee and I had reported and fixed issues in the mount propagation code.
+Each of these bugs had been hard to understand but only required trivial patches in order to be fixed.
+My expectation was no different for this bug.
+
+The terminology used in the mount propagation code used the now obsolete 'slave' and 'master' concepts to express dependency relationships.
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]:
```suggestion
The mount propagation code uses the now obsolete 'slave' and 'master' concepts to express dependency relationships.
```

Fix redundant use of the word "used."

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+
+* Slave mounts point to a peer in the closest peer group they receive propagation from via `@slave_mnt->mnt_master` (see above).
+  Together with these peers they form a propagation group (see below).
+  The closest peer group can thus be identified through the peer group id `@slave_mnt->mnt_master->mnt_group_id` of the peer/master that a slave mount receives propagation from.
+
+* A shared-slave mount is a slave mount to a peer group `pg1` while also a peer in another peer group `pg2`.
+  This simply means that a peer group may receive propagation from another peer group.
+
+  If a peer group `pg1` is a slave to another peer group `pg2` then all peers in peer group `pg1` point to the same peer in peer group `pg2` via `->mnt_master`.
+  So all peers in peer group `pg1` appear on the same `->mnt_slave_list`: they cannot be slaves to different peer groups.
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]: The previous paragraph refers to `pg1` as the master peer group and `pg2` as the slave group, then you switch that here. This paragraph then switches that around. I think this could be confusing; better to keep `pg1` as the master and `pg2` as the slave imo.

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+
+* A pure slave mount is a slave mount that is a slave to a peer group but is not a peer in another peer group.
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]:
```suggestion
* A pure slave mount is a slave to a peer group but is not a peer in another peer group.
```

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+  So a shared-slave mount that is a slave in `propg1` and that is a peer in another peer group `pg2` forms another propagation group `propg2` together with all slaves that point to that shared-slave mount in their `->mnt_master`.
+
+* A propagation tree refers to all mounts that receive propagation starting from a specific shared mount.
+
+  For example, for `propagate_mnt()` the destination mount `@dest_mnt` is the start of a propagation tree.
+  The propagation tree encompasses all mounts that receive propagation from `@dest_mnt`'s peer group down to the leafs.
+
+# The bug in one sentence
+
+The `propagate_mnt()` functions contains a bug where it fails to terminate at peers of `@source_mnt` when looking up copies of the source mount that become masters for copies of the source mount tree mounted on top of slaves in the destination propagation tree causing a NULL dereference.
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]: I find this sentence a little hard to parse, in large part because it's a difficult thing to summarize in one sentence. Maybe this is a little better?

```suggestion
The `propagate_mnt()` function contains a bug where it fails to terminate at peers of `@source_mnt` when searching for a copy of `@source_mnt` which is a suitable master for a new copy mounted on top of a slave in the destination propagation tree, causing a NULL dereference.
```

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 6]
```diff
@@ -0,0 +1,497 @@
...
+
+Hence, if we call into `propagate_one()` again for the next destination propagation node `@m`, `@last_dest` will point to the previous destination propagation node and `@last_source` will point to the previous copy of the source mount tree and mounted on `@last_dest`.
+
+Each new copy of the source mount tree is created from the previous copy of the source mount tree.
+This will become important later.
+
+The peer loop in `propagate_mnt()` is straightforward.
+We iterate through the peers copying and updating `@last_source` and `@last_dest` as we go through them and mount each copy of the source mount tree `@child` on a peer `@m` in `@dest_mnt`'s peer group.
+
+After `propagate_mnt()` handled the peers in `@dest_mnt`'s peer group `propagate_mnt()` will propagate the source mount tree down the propagation tree that `@dest_mnt`'s peer group propagates to:
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]:
```suggestion
After `propagate_mnt()` handles the peers in `@dest_mnt`'s peer group it will propagate the source mount down the propagation tree that `@dest_mnt`'s peer group propagates to:
```

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 7]
```diff
@@ -0,0 +1,497 @@
...
+                        goto out;
+                n = next_peer(n);
+        } while (n != m);
+}
+```
+
+The `next_group()` helper will recursively walk the destination propagation tree, descending into each propagation group of the propagation tree.
+
+The important part is that it takes care to propagate the source mount tree to all peers in the peer group of a propagation group before it propagates to the slaves to those peers in the propagation group.
+In other words, it creates and mounts copies of the source mount tree that become masters before it creates and mounts copies of the source mount tree that become slaves to these masters.
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]:
```suggestion
In other words, a mount in the source mount propagation tree which will be a master is always created before that mount's slaves.
```

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+
+It is important to remember that propagating the source mount tree to each mount `@m` in the destination propagation tree simply means that we create and mount new copies `@child` of the source mount tree on `@m` such that `@child->mnt_parent` points to `@m`.
+
+Since we know that each node `@m` in the destination propagation tree headed by `@dest_mnt`'s peer group will be overmounted with a copy of the source mount tree and since we know that the propagation properties of each copy of the source mount tree we create and mount at `@m` will mostly mirror the propagation properties of `@m`.
+We can use that information to create and mount the copies of the source mount tree that become masters before their slaves.
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]:
```suggestion
Since we know that each node `@m` in the destination propagation tree headed by `@dest_mnt`'s peer group will be overmounted with a copy of the source mount, and since we know that the propagation properties of each copy of the source mount tree we create and mount at `@m` will mostly mirror the propagation properties of `@m`, we can use that information to create and mount the copies of the source mount that become masters before their slaves.
```

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 9]
```diff
@@ -0,0 +1,497 @@
...
+
+The easy case is always when `@m` and `@last_dest` are peers in a peer group of a given propagation group.
+In that case we know that we can simply copy `@last_source` without having to figure out what the master for the new copy `@child` of the source mount tree needs to be as we've done that in a previous call to `propagate_one()`.
+
+The hard case is when we're dealing with a slave mount or a shared-slave mount `@m` in a destination propagation group that we need to create and mount a copy of the source mount tree on.
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]:
```suggestion
The hard case is when we're dealing with an `@m` which is a pure slave mount or a shared-slave mount in a new peer group, as we need to find an appropriate mount in the source mount tree to be the master of `@m`.
```

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 10]
```diff
@@ -0,0 +1,497 @@
...
+
+It is important to remember that propagating the source mount tree to each mount `@m` in the destination propagation tree simply means that we create and mount new copies `@child` of the source mount tree on `@m` such that `@child->mnt_parent` points to `@m`.
+
+Since we know that each node `@m` in the destination propagation tree headed by `@dest_mnt`'s peer group will be overmounted with a copy of the source mount tree and since we know that the propagation properties of each copy of the source mount tree we create and mount at `@m` will mostly mirror the propagation properties of `@m`.
+We can use that information to create and mount the copies of the source mount tree that become masters before their slaves.
+
+The easy case is always when `@m` and `@last_dest` are peers in a peer group of a given propagation group.
+In that case we know that we can simply copy `@last_source` without having to figure out what the master for the new copy `@child` of the source mount tree needs to be as we've done that in a previous call to `propagate_one()`.
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]:
```suggestion
In that case we know that the new copy `@child` should have the same master as `@last_source`, whose master was determined in a previous call to `propagate_one()`.
```

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 11]
```diff
@@ -0,0 +1,497 @@
...
+
+The hard case is when we're dealing with a slave mount or a shared-slave mount `@m` in a destination propagation group that we need to create and mount a copy of the source mount tree on.
+
+For each propagation group in the destination propagation tree we propagate the source mount tree to we want to make sure that the copies `@child` of the source mount tree we create and mount on slaves `@m` pick an earlier copy of the source mount tree that we mounted on a master `@m` of the destination propagation group as their master.
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]: I'm having a hard time with this sentence. Here's my attempt to put it in my own words:

```suggestion
For each pure slave or peer group in the destination propagation tree we need to make sure that the master for new copies of `@source_mnt` is a mount from source mount propagation tree whose parent is in the chain of masters of the parent for the new child mount.
```

One difference though is that I talk about a peer group and you talk about a propagation group. This is because `propagate_one()` searches for a new master based on `@m` and `@last_dest` not being peers, not based on your definition of a propagation group:

```c
        if (peers(m, last_dest)) {                               
                type = CL_MAKE_SHARED;
        } else {
            /* search for a new last_source with an appropriate mnt_master */
        }
```

So it's not making sense to me to talk about this on the basis of propagation groups.

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 12]
```diff
@@ -0,0 +1,497 @@
...
+This is a mouthful but as far as we can tell that's the core of it all.
+
+But, if we keep track of the masters in the destination propagation tree `@m` we can use the information to find the correct master for each copy of the source mount tree we create and mount at the slaves in the destination propagation tree `@m`.
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]: Here and in some other places you give a thing a name but never use the name for the thing. If you don't refer to the name then I think the name is unnecessary and just adds clutter.

```suggestion
But, if we keep track of the masters in the destination propagation tree `@m` we can use the information to find the correct master for each copy of the source mount tree we create and mount at the slaves in the destination propagation tree.
```

You probably also need to say something about how masters are marked, since afaict you haven't said that yet and you will start talking about looking for marked masters in a couple of sentences.

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 13]
```diff
@@ -0,0 +1,497 @@
...
+
+Let's walk through the base case as that's still fairly easy to grasp.
+
+If we're dealing with the first slave in the propagation group that `@dest_mnt` is in then we don't yet have marked any masters in the destination propagation tree.
+
+We know the master for the first slave to `@dest_mnt`'s peer group is simple `@dest_mnt`.
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]:
```suggestion
We know the master for the first slave to `@dest_mnt`'s peer group is simply `@dest_mnt`.
```

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 14]
```diff
@@ -0,0 +1,497 @@
...
+We walk up one more level since we don't have any marked mounts.
+So we end up with:
+
+```c
+n = dest_mnt;
+p = dest_mnt->mnt_master;
+```
+
+If `@dest_mnt`'s peer group is not slave to another peer group then `@p` is now `NULL`.
+If `@dest_mnt`'s peer group is a slave to another peer group then `@p` now points to `@dest_mnt->mnt_master` points which is a master outside the propagation tree we're dealing with.
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]:
```suggestion
If `@dest_mnt`'s peer group is a slave to another peer group then `@p` now points to `@dest_mnt->mnt_master`, which is a master outside the propagation tree we're dealing with.
```

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 15]
```diff
@@ -0,0 +1,497 @@
...
+        done = parent->mnt_master == p;
+        if (done && peers(n, parent))
+                break;
+        last_source = last_source->mnt_master;
+} while (!done);
+```
+
+We know that `@last_source->mnt_parent` points to `@last_dest` and `@last_dest` is the last peer in `@dest_mnt`'s peer group we propagated to in the peer loop in `propagate_mnt()`.
+
+Consequently, `@last_source` is the last copy we created and mount on that last peer in `@dest_mnt`'s peer group.
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]:
```suggestion
Consequently, `@last_source` is the last copy we created and mounted on that last peer in `@dest_mnt`'s peer group.
```

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 16]
```diff
@@ -0,0 +1,497 @@
...
+We also know that `@last_dest->mnt_master` is either `NULL` or points to a master outside of the destination propagation tree and so does `@p`.
+Hence:
+
+```c
+done = parent->mnt_master == p;
+```
+
+is trivially true in the base condition.
+
+We also know that for the first slave mount of `@dest_mnt`'s peer group that `@last_dest` either points `@dest_mnt` itself because it was initialized to:
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]:
```suggestion
We also know that for the first slave mount of `@dest_mnt`'s peer group, `@last_dest` either points `@dest_mnt` itself because it was initialized to:
```

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 17]
```diff
@@ -0,0 +1,497 @@
...
+```
+
+So, as we expected, we select `@last_source`, which refers to the last copy of the source mount tree we mounted on the last peer in `@dest_mnt`'s peer group, as the master of the first slave in `@dest_mnt`'s peer group.
+The rest is taken care of by `clone_mnt(last_source, ...)`.
+
+At the end of `propagate_mnt()` we now mark `@m->mnt_master` as the first master in the destination propagation tree that is distinct from `@dest_mnt->mnt_master`.
+Thus, we mark `@dest_mnt` itself as a master.
+
+By marking `@dest_mnt` or one of it's peers we are able to easily find it again when we later lookup masters for other copies of the source mount tree we mount copies of the source mount tree on slaves `@m` to `@dest_mnt`'s peer group.
+This, in turn allows us to find the master we selected for the copies of the source mount tree we mounted on master in the destination propagation tree again.
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]:
```suggestion
This in turn allows us to find the masters we selected for the copies of `@source_mnt`, which are always mounted on masters in the destination propagation tree.
```

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 18]
```diff
@@ -0,0 +1,497 @@
...
+What this means is that `@last_source` allows us to walk the destination propagation hierarchy the same way each destination propagation node `@m` does.
+
+If we take `@last_source`, which is the copy of `@source_mnt` we have mounted on `@last_dest` in the previous iteration of `propagate_one()`, then we know `@last_source->mnt_parent` points to `@last_dest` but we also know that as we walk through the destination propagation tree that `@last_source->mnt_master` will point to an earlier copy of the source mount tree we mounted one an earlier destination propagation node `@m`.
+
+So `@last_source->mnt_parent` will be our hook into the destination propagation tree and each consecutive `@last_source->mnt_master` will lead us to an earlier propagation node `@m` via `@last_source->mnt_master->mnt_parent`.
+
+Hence, by walking up `@last_source->mnt_master`, each of which is mounted on a node that is a master `@m` in the destination propagation tree we can also walk up the destination propagation hierarchy.
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]:
```suggestion
Hence, by walking up `@last_source->mnt_master`, each of which is mounted on a node that is a master in the destination propagation tree, we can also walk up the destination propagation hierarchy.
```

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 19]
```diff
@@ -0,0 +1,497 @@
...
+
+So, for each new destination propagation node `@m` we use the previous copy of `@last_source` and the fact it's mounted on the previous propagation node `@last_dest` via `@last_source->mnt_master->mnt_parent` to determine what the master of the new copy of `@last_source` needs to be.
+
+The goal is to find the _closest_ master that the new copy of the source mount tree we are about to create and mount on a slave `@m` in the destination propagation tree needs to pick.
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]: I don't understand what "needs to pick" means here.

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 20]
```diff
@@ -0,0 +1,497 @@
...
+This means we want to find a suitable master in the propagation group.
+
+As the propagation structure of the source mount propagation tree we create mirrors the propagation structure of the destination propagation
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]:
```suggestion
As the structure of the source mount propagation tree we create mirrors the propagation structure of the destination propagation
```

Using "propagation" twice here seems repetitive and unnecessary, and I think it reads easier like this.

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 21]
```diff
@@ -0,0 +1,497 @@
...
+tree we can find `@m`'s closest master - i.e., a marked master - which is a peer in the closest peer group that `@m` receives propagation from.
+We store that closest master of `@m` in `@p` as before and record the slave to that master in `@n`
+
+We then search for this master `@p` via `@last_source` by walking up the master hierarchy starting from the last copy of the source mount tree stored in `@last_source` that we created and mounted on the previous destination propagation node `@m`.
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]:
```suggestion
We then search for this master `@p` via `@last_source` by walking up the master hierarchy starting from `@last_source`.
```

You already explained what `@last_source` is so there's no need to explain it again here.

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 22]
```diff
@@ -0,0 +1,497 @@
...
+One scenario where this can happen is when we first handled a series of slaves of `@dest_mnt`'s peer group and then encounter peers in a new peer group that is a slave to `@dest_mnt`'s peer group.
+We handle them and then we encounter another slave mount to `@dest_mnt` that is a pure slave to `@dest_mnt`'s peer group.
+That pure slave will have a peer in `@dest_mnt`'s peer group as its master.
+Consequently, the new copy of the source mount tree will need to have `@source_mnt` as it's master.
+So we walk the propagation hierarchy all the way up to `@source_mnt` based on `@last_source->mnt_master`.
+
+So terminate on `@source_mnt`, easy peasy.
+Except, that the check misses something that the rest of the algorithm already handles.
+
+If `@dest_mnt` has peers in it's peer group the peer loop in `propagate_mnt()`:
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]:
```suggestion
If `@dest_mnt` has peers in its peer group the peer loop in `propagate_mnt()`:
```

File: _posts/2023-01-05-mount-propagation-bug.md [hunk 23]
```diff
@@ -0,0 +1,497 @@
...
+
+So the first time we handle a slave mount `@m` of `@dest_mnt`'s peer group the copy of the source mount tree we create will make the __last copy of the source mount tree we created and mounted on the last peer in `@dest_mnt`'s peer group the master of the new copy of the source mount tree we create and mount on the first slave of `@dest_mnt`'s peer group__.
+
+But this means that the termination condition that checks for `@source_mnt` is wrong.
+The `@source_mnt` cannot be found anymore by `propagate_one()`.
+Instead it will find the last copy of the source mount tree we created and mounted for the last peer of `@dest_mnt`'s peer group again.
+And that is a peer of `@source_mnt` not `@source_mnt` itself.
+
+This means, we fail to terminate the loop correctly and ultimately dereference `@last_source->mnt_master->mnt_parent`.
+When `@source_mnt`'s peer group isn't slave to another peer group then `@last_source->mnt_master` is `NULL` causing the splat below.
```
Comment (sforshee) [inline _posts/2023-01-05-mount-propagation-bug.md]:
```suggestion
When `@source_mnt`'s peer group isn't slave to another peer group then `@last_source->mnt_master` is `NULL` causing the splat above.
```

File: _posts/2023-01-05-mount-propagation-bug.md (hunk 3 above)
Comment (brauner) [reply _posts/2023-01-05-mount-propagation-bug.md]: Thank you, fixed!

File: _posts/2023-01-05-mount-propagation-bug.md (hunk 11 above)
Comment (brauner) [reply _posts/2023-01-05-mount-propagation-bug.md]: Yeah, I get what you mean.

File: _posts/2023-01-05-mount-propagation-bug.md (hunk 12 above)
Comment (brauner) [reply _posts/2023-01-05-mount-propagation-bug.md]: Thanks! Will do!

File: _posts/2023-01-05-mount-propagation-bug.md (hunk 19 above)
Comment (brauner) [reply _posts/2023-01-05-mount-propagation-bug.md]: Reformulated. Thanks!
